### DISCUSSION TITLE
Hydra does now allow to consent individual/partial scopes

### DISCUSSION THREAD

**Discussion by mrendi29** [ory/hydra#2300] [Q&A] [answered]
We use Hydra as our OAuth server and we are trying to allow the user to select which scopes they want to allow an application to use.
From the docs I can see that this behavior is supported as the docs show the use of checkboxes "https://www.ory.sh/hydra/docs/login-consent-flow/#the-flow-from-a-network-perspective".
When we try to accept the consent request with the list of scopes that the user wants to grant access to we get this error:

```
error: request_forbidden
error_description: The request is not allowed
```

You are not allowed to perform this action.
error_hint: You are not allowed to perform this action.
The following is also an example the response that Hydra gives after we try to accept this specific use case of consent_request:

`{'redirect_to': 'http://127.0.0.1:4444/oauth2/auth?client_id=43239a1a-eddd-4e70-21a8-12f9f90ec793&code_challenge=A5Rw4vdCSCcbv1rLUS6Cb2BQBDm-WIsAjZV5nJ1A3Tk&code_challenge_method=S256&consent_verifier=355b3d8daeaa848fdba568216ee391a2d&redirect_uri=http%3A%2F%2F127.0.0.1%3A3001%2Fauthorization&response_type=code&scope=great+great.ex3+great.ex2&state=kG8I25kWlUaE3HSLh3C6y6xGPiJ33m'}`

Does Hydra allow granting access only to partial scopes instead of all the requested scopes? If so, can that be documented more clearly, and if not, can we add that functionality as intended?

We are running on the latest Hydra version,  Flask for the back-end, and React for the front-end.

**Comment by vinckr** [comment, marked as answer]
I am not sure if I understood our problem correctly but did you set `grant_scope` correctly?
>When performing an OAuth 2.0 Flow where the end-user is involved (e.g. Implicit or Authorize Code), the granted OAuth 2.0 Scope must be set when accepting the consent using the grant_scope key. 

When I look at the `accept consent request` in the [reference API](https://www.ory.sh/hydra/docs/reference/api/#accept-a-consent-request) it only includes `grant_scope`.

Let me know if that was it, and where you think would be the best place to document this better. I am looking what else I can find on the scopes and `grant scope`.

**Reply by mrendi29** [reply]
Thank you for your reply. 

> I am not sure if I understood our problem correctly but did you set grant_scope correctly?

Yes. For ex: If an application were to request scopes: _A,B,C,D_ and the user only wants to grant access to scopes _B,C_ then these two scopes will be passed as the `grant_scope`, along with the `challenge`,`grant_access_token_audience`, `remember` and `remember_for` in the body when accepting the consent request.

**Reply by vinckr** [reply]
I see. 
And this example is what you are looking for, right? Just granting two scopes for example instead of all of them. 
This is definitely possible in Hydra, you can find an example in the 5 minute tutorial as well. 

Possibly an issue with your front end? How did you implement the consent endpoint?

**Reply by mrendi29** [reply]
Yes exactly. Allowing the user which scopes they want to allow instead of all of them. I noticed that this was possible on the 5 minute tutorial because i saw the checkboxes and I was surprised when i was getting that error.

In the front-end we  display all the scopes that the application requested and then the user can select which individual scopes they want to allow via checkboxes. The selected scopes are then passed via a form to the back-end along with the `challenge` and we use those selected scopes as the `grant_scope`. 

I have logged both the front-end and back-end for these values and they are passed and constructed correctly.

**Reply by vinckr** [reply]
Hey @mrendi29 ,
Sorry, it took a while.
Can you doublecheck that[ none of these apply](https://www.ory.sh/hydra/docs/debug/csrf/) ? I think your problem might be related to csrf.

**Reply by mrendi29** [reply]
Hi @vinckr,
We actually had not implemented CSRF yet. 
We are going to add it soon and I will update you on the outcome.

**Reply by vinckr** [reply]
Hey @mrendi29 
just checking in if your problem still persists or you are still implementing CSRF.

**Reply by mrendi29** [reply]
Hi Vincent, at the moment this feature is in our queue and will be soon
implemented. I will let you know as soon as we implement it whether the
error is sill persistent.
Thank you for checking in.

**Reply by mrendi29** [reply]
@vinckr Hi, This slipped out of my mind but writing to let you know that I figured out what was causing the issue. 
Even after implementing CSRF the issue was persisting. 
We were trying to accept the consent form from the front-end and then passing the response to the back end. 
Once we moved the consent flow to happen all in the back end this issue did not happen anymore :)

**Reply by vinckr** [reply]
Thanks for getting back to this Endi! 
Happy it got resolved, and maybe your experience can be useful to others 🙏
